types/ramda/test/propEq-tests.ts

```diff
@@ -10,7 +10,7 @@ import * as R from "ramda";
     const func1: (obj: Record<"foo", "bar">) => boolean = R.propEq("bar", "foo");
 
     const func2: (obj: Record<"baz", number>) => boolean = R.propEq(5, "baz");
-
+    // @ts-expect-error - key `5` not found on obj
```

Comment:
I was able to fix this and the test in filter-tests.ts by simplifying `propEq` to

```ts
export function propEq<T>(val: T): {
  // propEq(val)(name)(obj)
  <K extends PropertyKey>(name: K): (obj: Partial<Record<K, WidenLiterals<T>>>) => boolean;
  // propEq(val)(name, obj)
  <K extends PropertyKey>(name: K, obj: Partial<Record<K, WidenLiterals<T>>>): boolean;
};
// propEq(val, name)(obj)
export function propEq<const T, const K extends PropertyKey>(val: T, name: K): (obj: Partial<Record<K, WidenLiterals<T>>>) => boolean;
// propEq(val, name, obj)
export function propEq<K extends keyof U, U>(val: U[K], name: K, obj: U): boolean;
```

The type parameter `U` in the curried forms didn't seem to add much value and actually obfuscates the error you might get for an incompatible type since a failure only reports the argument isn't assignable to `never`.

While this may be slightly looser, it seems to work better higher-order as it allows `R.filter(isBuy, coll)` to check correctly.

Reply:
> The type parameter U in the curried forms didn't seem to add much value and actually obfuscates the error you might get for an incompatible type since a failure only reports the argument isn't assignable to never.

This was something we struggled with on that MR: https://github.com/ramda/types/pull/74/files#diff-5b720fabe3005a8d8170f7bdd36676339f3301612372fdb78c29189585317f0a

I put up an MR with your suggested change for comparison: https://github.com/ramda/types/pull/98

A bunch of tests start failing, let's examine one set of those tests:
```typescript
type Obj = {
  literals: 'A' | 'B';
  unions: number | string;
  nullable: number | null | undefined;
  optional?: number;
};

const obj = {} as Obj;

expectType<boolean>(propEq('1')('unions')(obj)); // fails 
expectType<boolean>(propEq('1', 'unions')(obj)); // fails
expectType<boolean>(propEq('1', 'unions', obj)); // ok
```
The first fails with
```
Argument of type Obj is not assignable to parameter of type Partial<Record<"unions", string>>.
  Types of property unions are incompatible.
    Type string | number is not assignable to type string | undefined.
      Type number is not assignable to type string.
```
Second with
```
Argument of type Obj is not assignable to parameter of type Partial<Record<"unions", string>>
```
The problem is the infered type for `T` here `'1'`, even after widened with `WidenLiterals<T>` to `string` is still to narrow for `string | number`.

What we are going for here is similar validation to the full signature in that `propEq<K extends keyof U, U>(val: U[K], name: K, obj: U)` can reject when `name` is not `keyof U` and `val` is not `U[K]`.

Because we know `U` here, we can typecheck in this way. In the curried versions, we have to do it in the opposite order. We try and reject `obj: U` if `U` doesn't have the key defined by `name: K` and if the type value at `U[K]` is `T`. This is surprisingly hard to do

In this case, the problem is that `string` in `Partial<Record<"unions", string>>` is not wide enough for `string | number` as defined for `type Obj` for key `unions`.

Because of this, and the many other failing tests, is why we had to go with the more complex solution.

> While this may be slightly looser, it seems to work better higher-order as it allows R.filter(isBuy, coll) to check correctly.

This was one I commented out and meant to came back to as well and forgot. While re-examining it for this comment, the changes made in  https://github.com/ramda/types/pull/74/files#diff-5b720fabe3005a8d8170f7bdd36676339f3301612372fdb78c29189585317f0a does affect how passing the result of `R.propEq("BUY", "type")` into `R.filter` that would break existing usage in a net negative way.

So in the end, I'm going to have to revert that MR and go back to the drawing board on it.